File: app/src/main/java/com/example/superps2/ui/LoadingActivity.java
package com.example.superps2.ui;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.example.superps2.R;
import com.example.superps2.PS2Bridge;

public class LoadingActivity extends Activity {
    private ProgressBar progressBar;
    private TextView statusText;
    private Handler handler;
    private String isoPath;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_loading);

        progressBar = findViewById(R.id.progressBar);
        statusText = findViewById(R.id.tvStatus);
        handler = new Handler(Looper.getMainLooper());

        // Get ISO path from intent
        isoPath = getIntent().getStringExtra("iso_path");
        if (isoPath == null) {
            finish();
            return;
        }

        // Start loading process
        new Thread(this::loadISO).start();
    }

    private void loadISO() {
        updateStatus("Loading ISO...");
        updateProgress(10);

        // Initialize emulator
        if (!PS2Bridge.init()) {
            updateStatus("Failed to initialize emulator");
            updateProgress(100);
            return;
        }
        updateProgress(30);

        // Load ISO
        if (!PS2Bridge.loadISO(isoPath)) {
            updateStatus("Failed to load ISO");
            updateProgress(100);
            return;
        }
        updateProgress(60);

        // Start emulator
        updateStatus("Starting emulator...");
        if (!PS2Bridge.runEmulator()) {
            updateStatus("Failed to start emulator");
            updateProgress(100);
            return;
        }
        updateProgress(90);

        // Success - launch game activity
        updateStatus("Starting game...");
        updateProgress(100);
        
        handler.postDelayed(() -> {
            Intent intent = new Intent(LoadingActivity.this, GameBootActivity.class);
            startActivity(intent);
            finish();
        }, 1000);
    }

    private void updateStatus(final String status) {
        handler.post(() -> statusText.setText(status));
    }

    private void updateProgress(final int progress) {
        handler.post(() -> progressBar.setProgress(progress));
    }
} 

File: app/src/main/java/com/superps2/emu/ui/LoadingActivity.kt
package com.superps2.emu.ui

import android.content.Intent
import android.os.Bundle
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.superps2.emu.jni.PS2Bridge
import com.superps2.R
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class LoadingActivity : AppCompatActivity() {
    private lateinit var progressBar: ProgressBar
    private lateinit var tvStatus: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_loading)

        progressBar = findViewById(R.id.progressBar)
        tvStatus = findViewById(R.id.tvStatus)

        // Recebe o path da ISO
        val isoPath = intent.getStringExtra("ISO_PATH")
        if (isoPath.isNullOrEmpty()) {
            Toast.makeText(this, "Nenhuma ISO fornecida", Toast.LENGTH_LONG).show()
            finish(); return
        }

        // Inicia o carregamento em background
        lifecycleScope.launch(Dispatchers.IO) {
            // Simular progresso incremental enquanto chama loadISO
            val loadJob = launch {
                val ok = PS2Bridge.startGame(isoPath)
                withContext(Dispatchers.Main) {
                    if (!ok) {
                        Toast.makeText(this@LoadingActivity, "Falha ao carregar ISO", Toast.LENGTH_LONG).show()
                        finish()
                    }
                }
            }

            // Enquanto carrega, animar a barra
            var progress = 0
            while (loadJob.isActive) {
                delay(100L)
                progress = (progress + 5).coerceAtMost(95)
                withContext(Dispatchers.Main) {
                    progressBar.progress = progress
                    tvStatus.text = "Carregando ISO... $progress%"
                }
            }
            // Aguarda término
            loadJob.join()
            // Finaliza progresso
            withContext(Dispatchers.Main) {
                progressBar.progress = 100
                tvStatus.text = "ISO carregada!"
            }
            delay(200L) // breve pausa para UX

            // Inicia GameBootActivity
            startActivity(Intent(this@LoadingActivity, GameBootActivity::class.java).apply {
                putExtra("ISO_PATH", isoPath)
            })
            overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
            finish()
        }
    }
} 